Userland/Libraries/LibWeb/MimeSniff/Resource.cpp

```diff
@@ -0,0 +1,625 @@
+    if (!any_of(resource.resource_header(), is_binary_data_byte))
+        return MimeType::create("text"_string, "plain"_string);
+
+    // 10. Return "application/octet-stream".
+    return MimeType::create("application"_string, "octet-stream"_string);
+}
+
+}
+
+namespace Web::MimeSniff {
+
+ErrorOr<Resource> Resource::create(ReadonlyBytes data, SniffingContext sniffing_context, bool no_sniff)
+{
+    // NOTE: Non-standard but for cases where pattern matching fails, let's fall back to the safest MIME type.
+    auto default_computed_mime_type = TRY(MimeType::create("application"_string, "octet-stream"_string));
+    auto resource = Resource { data, no_sniff, move(default_computed_mime_type) };
+
+    TRY(resource.supplied_mime_type_detection_algorithm(data));
+    TRY(resource.context_specific_sniffing_algorithm(sniffing_context));
+
+    return resource;
+}
+
+Resource::Resource(ReadonlyBytes data, bool no_sniff, MimeType&& default_computed_mime_type)
+    : m_no_sniff(no_sniff)
+    , m_computed_mime_type(move(default_computed_mime_type))
+{
+    read_the_resource_header(data);
+}
+
+Resource::Resource(Resource&&) = default;
+Resource& Resource::operator=(Resource&&) = default;
+
+Resource::~Resource() = default;
+
+// https://mimesniff.spec.whatwg.org/#supplied-mime-type-detection-algorithm
+ErrorOr<void> Resource::supplied_mime_type_detection_algorithm(ReadonlyBytes)
+{
+    // 1. Let supplied-type be null.
+    String supplied_type;
+
+    // FIXME: 2. If the resource is retrieved via HTTP, execute the following steps:
+    //        1. If one or more Content-Type headers are associated with the resource, execute the following steps:
+    //               1. Set supplied-type to the value of the last Content-Type header associated with the resource.
+    //               2. Set the check-for-apache-bug flag if supplied-type is exactly equal to one of the values in the following table:
+
+    // FIXME: 3. If the resource is retrieved directly from the file system, set supplied-type
+    //    to the MIME type provided by the file system.
+
+    // FIXME: 4. If the resource is retrieved via another protocol (such as FTP), set
+    //    supplied-type to the MIME type as determined by that protocol, if any.
+
+    // 5. If supplied-type is not a MIME type, the supplied MIME type is undefined.
+    //    Abort these steps.
+    // NOTE: supplied MIME type is undefined by default.
+    auto parsed_supplied_type = TRY(MimeType::parse(supplied_type));
+    if (!parsed_supplied_type.has_value())
+        return {};
+
+    // 6. The supplied MIME type is supplied-type.
+    m_supplied_mime_type = parsed_supplied_type.release_value();
+
+    return {};
+}
+
+// https://mimesniff.spec.whatwg.org/#read-the-resource-header
+void Resource::read_the_resource_header(ReadonlyBytes data)
+{
+    // 1. Let buffer be a byte sequence.
+    ByteBuffer buffer;
+
+    // 2. Read bytes of the resource into buffer until one of the following conditions is met:
```

Comment:
The problem is see here is that the byte buffer is assumed to be populated with all of the data. What the spec seems to be going for is kind of a streaming behavior, but the code here assumes that we have all of the data already loaded in memory. If we have the data all at once, extracting the header separately does not bring any additional benefits.

Reply:
Ah, now the mentions of "a reasonable amount of time has elapsed" and "slow connections" in the spec make sense. This sounds interesting, but I'm not quite sure how to approach this problem as I'm not sure if we have streaming API we can use to wait and consume resource data bytes until these conditions are met (and if the caller needs to account for this behavior too). I'll try to look around the code and find any types that would provide this streaming behavior

Reply:
I wanted to see how Firefox [handled this](https://searchfox.org/mozilla-central/source/netwerk/protocol/http/HttpTransactionChild.cpp#432) and it seems that they use their own Stream API, move the process of appending bytes to the buffer and checking if the max of 1445 bytes was reached into its own method, and has this `peekStream(...)` method that calls on this function (not sure how they handle reaching a certain elapsed time though)

Reply:
I think this may require more support for streams throughout the engine, especially the Fetch layer, which I believe only really provides complete buffers currently and doesn't use streams. For now, I think this is fine, but can be written with `slice`:
```cpp
static size_t MAX_SNIFF_SIZE = 1445;
buffer = data.slice(0, min(data.size(), MAX_SNIFF_SIZE));
```